packages/server/utils/OpenAIServerManager.ts

```diff
@@ -43,107 +41,101 @@ class OpenAIServerManager {
+        model: 'gpt-4',
+        messages: [
+          {
+            role: 'user',
+            content: prompt
+          }
+        ],
+        temperature: 0.7,
+        top_p: 1,
+        frequency_penalty: 0,
+        presence_penalty: 0
+      })
+      const themes = (response.data.choices[0]?.message?.content?.trim() as string) ?? null
+      return themes.split(', ')
+    } catch (e) {
+      const error = e instanceof Error ? e : new Error('OpenAI failed to generate themes')
+      console.error(error.message)
+      sendToSentry(error)
+      return null
+    }
   }
 
-  async groupReflections(reflectionsText: string[]) {
+  async groupReflections(reflectionsText: string[], themes: string[]) {
     if (!this.openAIApi) return null
-    const prompt = `You are given a list of reflections from a meeting, separated by commas. Your task is to group these reflections into topics and return an array of JavaScript objects. Each object represents a topic and contains an array of reflections that belong to that topic. Don't change the input text at all, even if it's spelt incorrectly or has special characters. Groups should have at least 2 reflections.
 
-    Example Input:
-    ['The retreat went well', 'The project might not be ready in time', 'The deadline is really tight', 'I liked that the length of the retreat', 'I enjoyed the hotel', 'Feeling overworked']
+    const getThemeForReflection = async (
+      reflection: string,
+      retry = false
+    ): Promise<string | null> => {
+      const prompt = `Given the themes ${themes.join(
+        ', '
+      )}, and the following reflection: "${reflection}", classify the reflection into the theme it fits in best. The reflection can only be added to one theme. Do not edit the reflection text. Your output should just be the theme name, and must be one of the themes I've provided.`
 
-    Example Output:
-    [
-    {
-    "The Retreat": [
-      "The retreat went well",
-      "I liked that the length of the retreat",
-      "I enjoyed the hotel"
-    ]
-    },
-    {
-    "Deadlines": [
-      "Feeling overworked",
-      "The project might not be ready in time",
-      "The deadline is really tight"
-    ]
-    }
-    ]
+      const response = await this.openAIApi!.createChatCompletion({
+        model: 'gpt-4',
+        messages: [
+          {
+            role: 'user',
+            content: prompt
+          }
+        ],
+        temperature: 0.7,
+        top_p: 1,
+        frequency_penalty: 0,
+        presence_penalty: 0
+      })
 
-    In the output, "The Retreat" and "Deadlines" are example topic names that you can create to group the reflections.
-    Do not give me code to run. Give me the output as valid JSON in the format above.
+      const theme = (response.data.choices[0]?.message?.content?.trim() as string) ?? null
+      if (!theme || !themes.includes(theme)) {
+        if (!retry) {
+          return getThemeForReflection(reflection, true)
```

Comment:
Sometimes, the model makes up a new theme, so I retry the request once more

Reply:
So it ignores this statement for some reason? Maybe somehow change it too? e.g. return the number of the theme

> Your output should just be the theme name, and must be one of the themes I've provided.

Reply:
Yeah, it ignores the explicit instruction 😄 I've tried a few different prompts, but hallucinating is a real problem with ChatGPT. Happy to test requesting the number of the theme, although my gut says that that would be harder for the model as it requires an extra step of looking for the index of the theme. Definitely worth testing though.